Do not consider min/max sizes when computing width as flex base size.

sammygill opened the pull request:
#### e093fe3f0975f00d1299770a0145d7d73a60219d
<pre>
Do not consider min/max sizes when computing width as flex base size.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244394">https://bugs.webkit.org/show_bug.cgi?id=244394</a>
rdar://99193252

Reviewed by Alan Bujtas.

According to the CSS flexbox spec, when we compute the base size for
flex items, we want to ignore the min and max main sizes of the item.
This is because these values will be used to compute the hypothetical
main size instead, The hypothetical main size will be computed using
the base size and clamping it in between the min and max main sizes.
This patch focuses specifically on the case when the main size is in
the inline axis (i.e. logical width).

In order to accomplish this, we need to keep track of when we are
computing base sizes in the flex algorithm. The m_isComputingBaseSizes
member will keep track of this and will be set/unset using the
SetForScope class. Then, whenever we compute the width of an item, we
will check to see if it is a flex item and if it is currently being
computed as a flex base size.

A little more work may be required to add the same type of logic to the
block axis. When attempting to do this in an initial implementation,
other test cases that started failing. It seems like that will require
an additional patch.

Spec: <a href="https://www.w3.org/TR/css-flexbox-1/#algo-main-item">https://www.w3.org/TR/css-flexbox-1/#algo-main-item</a>

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeReplacedLogicalWidthRespectingMinMaxWidth const):
(WebCore::RenderBox::shouldIgnoreMinMaxSizes const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexBaseSizeForChild):
* Source/WebCore/rendering/RenderFlexibleBox.h:
(WebCore::RenderFlexibleBox::isComputingBaseSizes const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computePreferredLogicalWidths):

Canonical link: <a href="https://commits.webkit.org/255002@main">https://commits.webkit.org/255002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/faf3292bc591132f0920b998330d30ee00869dec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21405 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100134 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158648 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33920 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28990 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83191 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96607 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/27025 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77647 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26830 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81782 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69861 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35006 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15549 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32807 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16531 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36583 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39481 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1530 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38508 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35620 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5617,4 +5620,11 @@ LayoutUnit synthesizedBaselineFromBorderBox(const RenderBox& box, LineDirectionM
 
+bool RenderBox::shouldIgnoreMinMaxSizes() const
+{
+    if (RenderElement *flexContainer = parent(); flexContainer && isFlexItem()) 
```
Seems like here testing isFlexItem() is enough since it implies the parent existing and being a flex container.

sammygill replied:
Good catch! Should be changed now. Since we still need the parent inside the if clause there's still a call the parent(), but this should reduce the number of calls when we don't ignore min-max sizes.

rwlbuis commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -668,6 +668,7 @@ override;
 
     bool shouldComputeLogicalHeightFromAspectRatio() const;
 
+    bool shouldIgnoreMinMaxSizes() const;
```
Style nit: please add an empty line after the method.

sammygill replied:
Fixed

rwlbuis commented on Source/WebCore/rendering/RenderFlexibleBox.h:
```diff
@@ -92,6 +92,7 @@ class RenderFlexibleBox : public RenderBlock {
 
     bool shouldApplyMinBlockSizeAutoForChild(const RenderBox&) const;
 
+    bool isComputingBaseSizes() const { return m_isComputingBaseSizes; }
```
Style nit: please add an empty line after the method.

sammygill replied:
Fixed

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5623,4 +5626,11 @@ LayoutUnit RenderBox::intrinsicLogicalWidth() const
+{
+    if (isFlexItem()) 
+        return downcast<RenderFlexibleBox>(parent())->isComputingFlexBaseSizes();
+    return false;
```
Have you considered return isFlexItem() && downcast<RenderFlexibleBox>(parent())->isComputingFlexBaseSizes();

sammygill committed:
Do not consider min/max sizes when computing width as flex base size.
https://bugs.webkit.org/show_bug.cgi?id=244394
rdar://99193252

Reviewed by Alan Bujtas.

According to the CSS flexbox spec, when we compute the base size for
flex items, we want to ignore the min and max main sizes of the item.
This is because these values will be used to compute the hypothetical
main size instead, The hypothetical main size will be computed using
the base size and clamping it in between the min and max main sizes.
This patch focuses specifically on the case when the main size is in
the inline axis (i.e. logical width).

In order to accomplish this, we need to keep track of when we are
computing base sizes in the flex algorithm. The m_isComputingBaseSizes
member will keep track of this and will be set/unset using the
SetForScope class. Then, whenever we compute the width of an item, we
will check to see if it is a flex item and if it is currently being
computed as a flex base size.

A little more work may be required to add the same type of logic to the
block axis. When attempting to do this in an initial implementation,
other test cases that started failing. It seems like that will require
an additional patch.

Spec: https://www.w3.org/TR/css-flexbox-1/#algo-main-item

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeReplacedLogicalWidthRespectingMinMaxWidth const):
(WebCore::RenderBox::shouldIgnoreMinMaxSizes const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexBaseSizeForChild):
* Source/WebCore/rendering/RenderFlexibleBox.h:
(WebCore::RenderFlexibleBox::isComputingBaseSizes const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computePreferredLogicalWidths):

Canonical link: https://commits.webkit.org/255002@main
